enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/upstream/UpstreamDatabaseStrategySelector.java

```diff
@@ -59,30 +61,28 @@ public UpstreamDatabaseStrategySelector( UpstreamDatabaseSelectionStrategy defau
 
     public MemberId bestUpstreamDatabase() throws UpstreamDatabaseSelectionException
     {
-        MemberId result = null;
-        for ( UpstreamDatabaseSelectionStrategy strategy : strategies )
+        MemberId result = strategies.stream()
+                .map( strategyMemberId() )
+                .filter( Optional::isPresent )
+                .map( Optional::get )
+                .findFirst()
+                .orElseThrow( () -> new UpstreamDatabaseSelectionException( "Could not find an upstream database with which to connect." ) );
+        log.debug( "Selected upstream database [%s]", result );
+        return result;
+    }
+
+    private static Function<UpstreamDatabaseSelectionStrategy,Optional<MemberId>> strategyMemberId()
+    {
+        return strategy ->
         {
-            log.debug( "Trying selection strategy [%s]", strategy.toString() );
             try
             {
-                if ( strategy.upstreamDatabase().isPresent() )
-                {
-                    result = strategy.upstreamDatabase().get();
-                    break;
-                }
+                return strategy.upstreamDatabase();
             }
-            catch ( NoSuchElementException ex )
+            catch ( NoSuchElementException | UpstreamDatabaseSelectionException e )
```

Comment:
This hides the actual cause of the UpstreamDatabaseSelectionException

Reply:
Reverted